src/app/clusters/thread-network-directory-server/thread-network-directory-server.cpp

```diff
@@ -0,0 +1,288 @@
+}
+
+CHIP_ERROR ThreadNetworkDirectoryServer::ReadThreadNetworks(const ConcreteDataAttributePath &, AttributeValueEncoder & aEncoder)
+{
+    return aEncoder.EncodeList([this](const auto & encoder) -> CHIP_ERROR {
+        CHIP_ERROR err = CHIP_NO_ERROR;
+        ExtendedPanId exPanId;
+        auto * iterator = mStorage.IterateNetworkIds();
+        while (iterator->Next(exPanId))
+        {
+            uint8_t datasetBuffer[kSizeOperationalDataset];
+            MutableByteSpan datasetSpan(datasetBuffer);
+            SuccessOrExit(err = mStorage.GetNetworkDataset(exPanId, datasetSpan));
+
+            OperationalDataset dataset;
+            char networkName[kSizeNetworkName + 1];
+            ThreadNetworkStruct::Type network;
+
+            dataset.Init(datasetSpan);
+            SuccessOrExit(err = dataset.GetExtendedPanIdAsByteSpan(network.extendedPanID));
+            SuccessOrExit(err = dataset.GetNetworkName(networkName));
+            network.networkName = CharSpan::fromCharString(networkName);
+            SuccessOrExit(err = dataset.GetChannel(network.channel));
+            SuccessOrExit(err = dataset.GetActiveTimestamp(network.activeTimestamp));
+
+            SuccessOrExit(err = encoder.Encode(network));
+        }
+    exit:
+        iterator->Release();
+        return err;
+    });
+}
+
+void ThreadNetworkDirectoryServer::InvokeCommand(HandlerContext & ctx)
+{
+    switch (ctx.mRequestPath.mCommandId)
+    {
+    case AddNetwork::Id:
+        HandleCommand<AddNetwork::DecodableType>(
+            ctx, [this](HandlerContext & aCtx, const auto & req) { HandleAddNetworkRequest(aCtx, req); });
+        return;
+    case RemoveNetwork::Id:
+        HandleCommand<RemoveNetwork::DecodableType>(
+            ctx, [this](HandlerContext & aCtx, const auto & req) { HandleRemoveNetworkRequest(aCtx, req); });
+        return;
+    case GetOperationalDataset::Id:
+        HandleCommand<GetOperationalDataset::DecodableType>(
+            ctx, [this](HandlerContext & aCtx, const auto & req) { HandleOperationalDatasetRequest(aCtx, req); });
+        return;
+    }
+}
+
+void ThreadNetworkDirectoryServer::HandleAddNetworkRequest(HandlerContext & ctx,
+                                                           const ThreadNetworkDirectory::Commands::AddNetwork::DecodableType & req)
+{
+    OperationalDataset dataset;
+    ByteSpan extendedPanIdSpan;
+    union
+    {
+        uint64_t activeTimestamp;
+        uint16_t channel;
+        uint8_t masterKey[kSizeMasterKey];
+        uint8_t meshLocalPrefix[kSizeMeshLocalPrefix];
+        char networkName[kSizeNetworkName + 1];
+        uint16_t panId;
+        uint8_t pksc[kSizePSKc];
+        uint32_t securityPolicy;
+    } unused;
+    ByteSpan unusedSpan;
+
+    // "It SHALL contain at least the following sub-TLVs: Active Timestamp, Channel, Channel Mask,
+    // Extended PAN ID, Network Key, Network Mesh-Local Prefix, Network Name, PAN ID, PKSc, and Security Policy."
```

Comment:
Followup: should this validation method live on OperationalDataset?

Reply:
I guess that would avoid having to copy out all the values into dummy fields too. I'm torn about whether we should be validating that these elements have the right sizes (where they are fixed size)... the current code does, but we also want the Thread Network Directory to be forward compatible with new Thread versions (though I guess one could reasonably expect them to not change the size of TLV elements that are defined as having a fixed size in the current spec)

Reply:
I'm leaving it as is for now, will keep it in mind for a follow-up